pkg/engine/lifecycletest/refresh_before_update_test.go

```diff
@@ -0,0 +1,151 @@
+						RefreshBeforeUpdate: true,
+					}, nil
+				},
+				DiffF: func(_ context.Context, req plugin.DiffRequest) (plugin.DiffResponse, error) {
+					if req.NewInputs.DeepEquals(req.OldInputs) {
+						return plugin.DiffResponse{Changes: plugin.DiffNone}, nil
+					}
+					return plugin.DiffResponse{Changes: plugin.DiffSome}, nil
+				},
+				UpdateF: func(_ context.Context, req plugin.UpdateRequest) (plugin.UpdateResponse, error) {
+					assert.Equal(t, "<FRESH-INPUT>", req.OldInputs["input"].StringValue())
+					assert.Equal(t, "<FRESH-RESULT>", req.OldOutputs["result"].StringValue())
+
+					props := req.NewInputs.Copy()
+					props["result"] = props["input"]
+					return plugin.UpdateResponse{
+						Properties:          props,
+						RefreshBeforeUpdate: true,
+					}, nil
+				},
+				ReadF: func(_ context.Context, req plugin.ReadRequest) (plugin.ReadResponse, error) {
+					inputs := req.Inputs.Copy()
+					inputs["input"] = resource.NewStringProperty("<FRESH-INPUT>")
+					props := req.State.Copy()
+					props["result"] = resource.NewStringProperty("<FRESH-RESULT>")
+					return plugin.ReadResponse{
+						Status: resource.StatusOK,
+						ReadResult: plugin.ReadResult{
+							ID:                  "new-id",
+							Inputs:              inputs,
+							Outputs:             props,
+							RefreshBeforeUpdate: true,
+						},
+					}, nil
+				},
+			}, nil
+		}),
+	}
+
+	inputs := resource.PropertyMap{
+		"input": resource.NewStringProperty("value-1"),
+	}
+
+	programF := deploytest.NewLanguageRuntimeF(func(_ plugin.RunInfo, monitor *deploytest.ResourceMonitor) error {
+		_, err := monitor.RegisterResource("pkgA:m:typA", "resA", true, deploytest.ResourceOptions{
+			Inputs: inputs,
+		})
+		assert.NoError(t, err)
+
+		return nil
+	})
+	hostF := deploytest.NewPluginHostF(nil, nil, programF, loaders...)
+
+	p := &lt.TestPlan{
+		Options: lt.TestUpdateOptions{
+			T:     t,
+			HostF: hostF,
+		},
+	}
+
+	// First update.
+	p.Steps = []lt.TestStep{{Op: Update}}
+	snap := p.Run(t, nil)
+	assert.Len(t, snap.Resources, 2)
+	assert.Equal(t, "default", snap.Resources[0].URN.Name())
+	assert.Equal(t, "resA", snap.Resources[1].URN.Name())
+	assert.Equal(t, "value-1", snap.Resources[1].Inputs["input"].StringValue())
+	assert.Equal(t, "value-1", snap.Resources[1].Outputs["result"].StringValue())
+	assert.True(t, snap.Resources[1].RefreshBeforeUpdate)
+
+	// Second update.
+	inputs = resource.PropertyMap{
+		"input": resource.NewStringProperty("value-2"),
+	}
+	p.Steps = []lt.TestStep{{Op: Update}}
+	snap = p.Run(t, snap)
+	assert.Len(t, snap.Resources, 2)
+	assert.Equal(t, "default", snap.Resources[0].URN.Name())
+	assert.Equal(t, "resA", snap.Resources[1].URN.Name())
+	assert.Equal(t, "value-2", snap.Resources[1].Inputs["input"].StringValue())
+	assert.Equal(t, "value-2", snap.Resources[1].Outputs["result"].StringValue())
+	assert.True(t, snap.Resources[1].RefreshBeforeUpdate)
+
+	// Fresh import with new snapshot.
```

Comment:
No need to separately test `pulumi refresh`-like scenario separately from `pulumi import`-like scenario here?

Reply:
I'll add one.